src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs

```diff
@@ -88,6 +91,224 @@ public void TestAsyncOutputStream()
             }
         }
 
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap, "Pipe doesn't work well on UAP")]
+        async public Task TestAsyncOutputStream_CancelOutputRead()
+        {
+            using (AnonymousPipeServerStream pipeWrite = new AnonymousPipeServerStream(PipeDirection.Out, HandleInheritability.Inheritable))
+            using (AnonymousPipeServerStream pipeRead = new AnonymousPipeServerStream(PipeDirection.In, HandleInheritability.Inheritable))
+            {
+                using (Process p = CreateProcess(TestAsyncOutputStream_CancelOutputRead_RemotelyInvokable, $"{pipeWrite.GetClientHandleAsString()} {pipeRead.GetClientHandleAsString()}"))
+                {
+                    var dataReceived = new List<int>();
+                    var dataArrivedEvent = new AutoResetEvent(false);
+
+                    p.StartInfo.RedirectStandardOutput = true;
+                    p.OutputDataReceived += (s, e) =>
+                    {
+                        if (e.Data != null)
+                        {
+                            dataReceived.Add(int.Parse(e.Data));
+                        }
+                        dataArrivedEvent.Set();
+                    };
+
+                    // Start child process
+                    p.Start();
+
+                    pipeWrite.DisposeLocalCopyOfClientHandle();
+                    pipeRead.DisposeLocalCopyOfClientHandle();
+
+                    // Wait child process start
+                    Assert.True(await WaitPipeSignal(pipeRead, WaitInMS), "Child process not started");
+
+                    //Start listening and produce output 1
+                    p.BeginOutputReadLine();
+                    await pipeWrite.WriteAsync(new byte[1], 0, 1);
+
+                    // Wait child signal produce number 1
+                    Assert.True(await WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 1");
+                    Assert.True(dataArrivedEvent.WaitOne(WaitInMS), "Value 1 not received");
+
+                    // Stop listening and signal to produce value 2
+                    p.CancelOutputRead();
+                    await pipeWrite.WriteAsync(new byte[1], 0, 1);
+
+                    // Wait child signal produce number 2
+                    Assert.True(await WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 2");
+
+                    // Wait child process close to be sure that output buffer has been flushed
+                    Assert.True(p.WaitForExit(WaitInMS), "Child process didn't close");
```

Comment:
> The former would synchronize between the parent and child by having the child send data and the parent read it all, then canceling and then signaling to the child to send more, then the child exiting, and the parent validating that it saw the stream end without having received additional data from the child.

@stephentoub I'm not sure about this, what if child exits but signal 2 already arrived on parent but there a delay in "parent side processing", we could receive data also after asserts, right?

Reply:
ping @stephentoub

Reply:
Hmm.  I guess the key issue is that there's no API exposed that let's the test "see the stream end", as I suggested.  I'll need to think about it.  In the meantime, I don't want to hold up the product fix.  Why don't you just add a comment that there's a race condition such that the test might have some false negatives.

Reply:
I think false positives, because we cancel output and ask to child to send send new one, after that we wait for process exit and count dataReceived(expected unchanged), but if we've a bug in implementation the message could be "in processing" on parent side but we passed tests. BTW added comments.

Reply:
I think we're saying the same thing but using opposite polarity. Just to be sure, when I say it may have false negatives, I mean the test may fail to catch a bug (to me a false positive is the test saying there's a bug when there isn't one).

Reply:
ah ok :smile: I'll change the polarity on comment